### PR TITLE
Add COMAnalyzer

### PR DESCRIPTION
**Pull request by mateoatr**
This adds support for `IL2050`. It is intended to be used for all other diagnostics related to COM in the analyzers. I didn't add any new tests -- we are reusing the ones under `Interop`.

### CODE REVIEW

File: src/ILLink.RoslynAnalyzer/COMAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+					typedConstant.Type != null && typedConstant.Type.IsUnmanagedType) {
+					var unmanagedType = typedConstant.Value;
+					switch (unmanagedType) {
+					case (int) UnmanagedType.IUnknown:
+					case (int) UnmanagedType.IDispatch:
+					case (int) UnmanagedType.Interface:
+						return true;
+
+					default:
+						break;
```
Comment (sbomer) [inline src/ILLink.RoslynAnalyzer/COMAnalyzer.cs:68]: The linker's logic has an explicit `nativeType == NativeType.None` check - do we need something like that? (Is there a case where the marshal info can change any of the below `true` returns to be not COM related?)

File: src/ILLink.RoslynAnalyzer/COMAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+					}
+				}
+
+				if (IsInterface (symbol))
+					return true;
+
+				if (symbol is not IParameterSymbol parameterSymbol)
```
Comment (sbomer) [inline src/ILLink.RoslynAnalyzer/COMAnalyzer.cs:75]: Don't we need to consider the below cases also for return types? Or are those also represented as `IParameterSymbol`?

File: src/ILLink.RoslynAnalyzer/COMAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+				} else if (IsInterface (parameterSymbol.Type)) {
+					// Interface types marshal as COM by default
+					return true;
+				} else if (parameterType.ContainingNamespace.Name == "System" &&
+					parameterType.Name == "MulticastDelegate") {
+					// Delegates are special cased by interop
+					return false;
+				} else if (parameterType.ContainingNamespace.Name == "System.Runtime.InteropServices" &&
+					(parameterType.Name == "CriticalHandle" || parameterType.Name == "SafeHandle")) {
+					// Subclasses of CriticalHandle and SafeHandle are special cased by interop
```
Comment (sbomer) [inline src/ILLink.RoslynAnalyzer/COMAnalyzer.cs:100]: Do we need to check for subclasses here?

File: test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs [hunk 4]
```diff
@@ -83,12 +85,16 @@ public static void RunTest<TAnalyzer> (MemberDeclarationSyntax m, List<Attribute
 			var builder = ImmutableDictionary.CreateBuilder<string, List<string>> ();
 
 			foreach (var file in GetTestFiles ()) {
-				var dirName = Path.GetFileName (Path.GetDirectoryName (file))!;
-				if (builder.TryGetValue (dirName, out var sources)) {
+				var dirName = Path.GetDirectoryName (file);
```
Comment (sbomer) [inline test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs:88]: nit: Might just be me, but I find it confusing that `GetDirectoryName` gives back the directory path, not just the name. Would you mind renaming things to make this a bit more readable? For example `dirName` -> `directory`, maybe introduce a `parentDirectory` variable?

**Review by sbomer — COMMENTED** [review]
Nice!

File: src/ILLink.RoslynAnalyzer/COMAnalyzer.cs (hunk 1 above)
Comment (mateoatr) [reply src/ILLink.RoslynAnalyzer/COMAnalyzer.cs:70]: I added a check to see if the typed constant value exists in `UnmanagedType` (which mimics the check of `nativeType == NativeType.None` in Cecil).

File: src/ILLink.RoslynAnalyzer/COMAnalyzer.cs (hunk 3 above)
Comment (mateoatr) [reply src/ILLink.RoslynAnalyzer/COMAnalyzer.cs:105]: Yup. Added a helper method in ISymbolExtensions for this. I moved the `IsInterface` method there as well.

**Review by sbomer — APPROVED** [review]
Thank you, LGTM!

**Comment by marek-safar** [comment]
@mateoatr I had to revert this change as it caused test failures, please re-submit it after rebasing.

**Pull request merged**